### PR TITLE
nm: Activate new virtual interfaces

### PR DESCRIPTION
**Pull request by EdDev**
NMState supports several virtual interfaces, all having a master-slave
relationship (bond, ovs-bridge). The master interface has been activated
so far by its slaves (which automatically perform such an activation
when they are activated).
With the introduction of VLAN interfaces, it is now required for the
virtual interface to be explicitly activated.

Activating a new interface is now supported by executing an activation
command on the (remote) connection profile. Existing interfaces are
still activated using their device object.

### CODE REVIEW

File: libnmstate/nm/applier.py [hunk 1]
```diff
@@ -85,23 +86,73 @@ def prepare_edited_ifaces_configuration(ifaces_desired_state):
...
+    """
+    Control interface admin state by activating, deactivating and deleting
+    devices connection profiles.
+
+    The `down` and `absent` states are handled in the same manner (temporary)
+    by deactivating the device and deleting the connection profile.
+
+    For new virtual devices, the `up` state is handled by activating the
+    new connection profile. For existing devices, the device is activated,
+    leaving it to choose the correct profile.
```
Comment (tyll) [inline libnmstate/nm/applier.py]: What is the advantage of activating the device instead of always activating the profile that we want?

File: libnmstate/nm/applier.py [hunk 2]
```diff
@@ -85,23 +86,73 @@ def prepare_edited_ifaces_configuration(ifaces_desired_state):
     return con_profiles
 
 
-def set_ifaces_admin_state(ifaces_desired_state):
+def set_ifaces_admin_state(ifaces_desired_state, con_profiles=()):
+    """
+    Control interface admin state by activating, deactivating and deleting
+    devices connection profiles.
+
+    The `down` and `absent` states are handled in the same manner (temporary)
```
Comment (tyll) [inline libnmstate/nm/applier.py]: How about:
```
The `absent` state results in deactivating the device and deleting the connection profile.
FIXME: The `down` state is currently handled in the same way.
```

This makes it more clear that the down state may change in the future.

File: libnmstate/nm/applier.py [hunk 3]
```diff
@@ -85,23 +86,73 @@ def prepare_edited_ifaces_configuration(ifaces_desired_state):
...
+    by deactivating the device and deleting the connection profile.
+
+    For new virtual devices, the `up` state is handled by activating the
+    new connection profile. For existing devices, the device is activated,
+    leaving it to choose the correct profile.
+
+    In order to activate correctly the interfaces, the order is significant:
+    - New interfaces (virtual interfaces).
```
Comment (tyll) [inline libnmstate/nm/applier.py]: Why does it need to be this order? What happens when we want to use a VLAN on top of a bond interface, doesn't the bond interface need to be activated first?

File: libnmstate/nm/device.py [hunk 4]
```diff
@@ -20,13 +20,28 @@
 from . import nmclient
 
 
-def activate(dev, connection=None):
+def activate(dev, fetch_remote_connection_func=None):
+    """
+    Activating the given device or remote connection profile.
```
Comment (tyll) [inline libnmstate/nm/device.py]: minor detail: Better start the description in imperative mood, i.e. `Activate the given device... `

File: libnmstate/nm/device.py [hunk 5]
```diff
@@ -20,13 +20,28 @@
 from . import nmclient
 
 
-def activate(dev, connection=None):
+def activate(dev, fetch_remote_connection_func=None):
```
Comment (tyll) [inline libnmstate/nm/device.py]: Shouldn't the dev parameter be optional?

File: tests/integration/nm/vlan_test.py [hunk 6]
```diff
@@ -35,7 +35,7 @@ def test_create_and_remove_vlan():
         assert vlan_desired_state == vlan_current_state
 
     # FIXME: Remove this sleep after adding wait for state mechanism.
-    time.sleep(2)
+    time.sleep(5)
```
Comment (tyll) [inline tests/integration/nm/vlan_test.py]: This should probably be a separate PR/commit, since it is unrelated/fixes problems for other PRs as well.

File: tests/integration/nm/vlan_test.py [hunk 7]
```diff
@@ -67,9 +67,10 @@ def _create_vlan(vlan_desired_state):
     with mainloop():
         con_profile = nm.connection.create_profile((con_setting, vlan_setting))
         nm.connection.add_profile(con_profile, save_to_disk=False)
-    with mainloop():
-        remote_profile = nm.nmclient.client().get_connection_by_id(ifname)
-        nm.device.activate(dev=None, connection=remote_profile)
+        nm.device.activate(
+            dev=None,
+            fetch_remote_connection_func=(
+                lambda: nm.nmclient.client().get_connection_by_id(ifname)))
```
Comment (tyll) [inline tests/integration/nm/vlan_test.py]: It seems that `activate()` could be easily be simplified by just accepting a connection id instead of the fetch function.

File: tests/lib/nm/device_test.py [hunk 8]
```diff
@@ -41,10 +41,11 @@ def mainloop_mock():
 
 def test_activate(client_mock, mainloop_mock):
     dev = 'foodev'
-    nm.device.activate(dev)
+    mainloop_mock.push_action = lambda func, dev, fetch: func(dev, fetch)
 
-    mainloop_mock.push_action.assert_called_once_with(
-        client_mock.activate_connection_async,
+    nm.device.activate(dev, fetch_remote_connection_func=None)
```
Comment (tyll) [inline tests/lib/nm/device_test.py]: The fetch function is by default None, why specify it explicitly here? It seems that there is a test missing in case it is not None.

File: libnmstate/nm/applier.py [hunk 9]
```diff
@@ -85,23 +86,73 @@ def prepare_edited_ifaces_configuration(ifaces_desired_state):
...
+                for nmdev in nmdevs:
+                    devs_actions[nmdev] = (device.deactivate, device.delete)
+            else:
+                raise UnsupportedIfaceStateError(iface_desired_state)
+
+    for ifname in new_ifaces_to_activate:
+        device.activate(
+            dev=None,
+            fetch_remote_connection_func=(
+                lambda: nmclient.client().get_connection_by_id(ifname))
```
Comment (tyll) [inline libnmstate/nm/applier.py]: It seems that activate() could be easily be simplified by just accepting a connection id instead of the fetch function.

File: libnmstate/nm/applier.py (hunk 3 above)
Comment (EdDev) [reply libnmstate/nm/applier.py]: Good point, we need new VLAN/s last in the 'new' interfaces section.
It's actually even more complex with bridges and friends.

The order matters only for masters before their slaves. Tests failed when activating the slave before the master, as the master explicit activation collided with the slaves triggering the same implicitly.
This may become a problem with the introduction of other device types and more scenarios.

File: libnmstate/nm/applier.py (hunk 9 above)
Comment (EdDev) [reply libnmstate/nm/applier.py]: The function is triggered only later by the mainloop, there is no connection ID at this time, it has not been created yet (the connection is created in the mainloop as well).

File: libnmstate/nm/device.py (hunk 5 above)
Comment (EdDev) [reply libnmstate/nm/device.py]: It can be. I'll change it.

File: libnmstate/nm/device.py (hunk 4 above)
Comment (EdDev) [reply libnmstate/nm/device.py]: Done

File: tests/integration/nm/vlan_test.py (hunk 6 above)
Comment (EdDev) [reply tests/integration/nm/vlan_test.py]: Right

File: tests/integration/nm/vlan_test.py (hunk 7 above)
Comment (EdDev) [reply tests/integration/nm/vlan_test.py]: see previous answer to this

File: tests/lib/nm/device_test.py (hunk 8 above)
Comment (EdDev) [reply tests/lib/nm/device_test.py]: Done

Comment (EdDev) [reply tests/lib/nm/device_test.py]: Regarding the missing test: It is covered by the nm.vlan test in the integration suite.
I will start taking out these unit nm tests in future patches.

**Comment by EdDev** [comment]
ci test please

**Comment by EdDev** [comment]
ci test please

**Comment by EdDev** [comment]
ci test please

File: libnmstate/nm/applier.py (hunk 9 above)
Comment (EdDev) [reply libnmstate/nm/applier.py]: > his might true for now, but there is no document of NM confirmed this.

It is document in our code:
https://github.com/nmstate/nmstate/blob/master/libnmstate/nm/connection.py#L105

> How about use nmclient.activate_connection_async where you can define device without connection/profile ID?

I think nmstate is supposed to be specific when controlling the interfaces. I am not clear how we could do that without a unique identifier (we have today the device and the profile ID).
But there may be a need to add other identifiers in the future, if and when that  happens we could add it like we added this one.

File: libnmstate/nm/applier.py (hunk 3 above)
Comment (tyll) [reply libnmstate/nm/applier.py]: It does not need to happen in this PR, but could we instead just check the errors in case of collisions and ignore errors if it just says that we want to activate an already activated connection?

File: libnmstate/nm/applier.py (hunk 1 above)
Comment (tyll) [reply libnmstate/nm/applier.py]: I can agree with the first point. The second point seems to be wrong because the best profile will always be the one that nmstate created - otherwise we will not reach the defined state in nmstate. There is no clear argument for the third one since it seems to be more of a gut feeling.

File: libnmstate/nm/applier.py (hunk 9 above)
Comment (EdDev) [reply libnmstate/nm/applier.py]: Ahh, you are absolutely right. I was blind.
Fixed! Thank you!

File: libnmstate/nm/device.py [hunk 10]
```diff
@@ -20,13 +20,28 @@
 from . import nmclient
 
 
-def activate(dev, connection=None):
+def activate(dev=None, connection_id=None):
+    """
+    Activate the given device or remote connection profile.
+    For the remote connection profile, a fetcher function is provided which
```
Comment (tyll) [inline libnmstate/nm/device.py]: The docstring needs to be adjusted for the change from fetch function to connection_id

Comment (EdDev) [reply libnmstate/nm/device.py]: Done

**Comment by EdDev** [comment]
ci test please

File: libnmstate/nm/device.py [hunk 11]
```diff
@@ -20,13 +20,25 @@
 from . import nmclient
 
 
-def activate(dev, connection=None):
+def activate(dev=None, connection_id=None):
+    """
+    Activate the given device or remote connection profile.
+    For the remote connection profile, the connection ID is provided.
```
[the remark this answers is not in the document]
Comment (EdDev) [reply libnmstate/nm/device.py]: Fair enough, removing.
